### PR TITLE
fix(list-box): ensure size variants are in effect

### LINKED ISSUE #4916: [Dropdown] xl size doesn't work
The height is overridden by the default dropdown height.

reproduction: https://carbon-components-react.netlify.com/?path=/story/dropdown--default

1. activate the `xl` knob (no effect)
2. open dev tools and un-check the height coming from default dropdown styles

### PR DESCRIPTION
**Pull request by asudoh**
This change ensures the style rule in `.bx--list-box--xl` for dropdown size wins over one in `.bx--dropdown` or any ones in application styles (for default size), by expressing that they are for non-default styles and should win order the one for the default size.

Fixes #4916.

#### Changelog

**Changed**

- Adding `.bx--list-box` to the selector of size variant style rule, so it wins over the style rule for the default height.

#### Testing / Reviewing

Testing should make sure combo box, dropdown, multi select are not broken.

### CODE REVIEW

**Review by vpicone — APPROVED** [review]

**Review by vpicone — DISMISSED** [review]
@asudoh I don't believe this is the correct approach. We don't want double the specificity for all listbox components it makes it very difficult to override and theme. Say for instance someone wanted to make the 'large' button larger in their theme. This would be unexpected specificity and require `important` or double/triple classNames.

This is also different than how we do variants in the codebase. Take `Button` for example: the variants (size and kind) don't use double class names to style.

**Comment by vpicone** [comment]
Also wondering if the dropdown options should be the same size as the field. @carbon-design-system/design 

Screen Shot 2020-01-03 at 3 52 07 PM

**Comment by mjabbink** [comment]
yes. The dropdown and fields below should be same. so for 32px dd then 32px for content, and for 40 and 48px options the same I think. @jeanservaas?

**Comment by asudoh** [comment]
Wrt visual design it makes sense. Given it does not seem to be in the scope of #4916, created #4951 to track that.

Wrt the CSS rule, I see the point of difference from buttons, while I think it's a sensible difference given where we are at. Just in case - Would you want to suggest an alternate approach @vpicone?

**Comment by vpicone** [comment]
@asudoh yes, creating an exception for non-listbox implementations (vanilla) as I did in my original PR would both prevent excess specificity and allow us to use the strategy found in other variant implementations (such as button).

**Comment by asudoh** [comment]
Refs: https://github.com/carbon-design-system/carbon/pull/4917#issuecomment-567263864

**Comment by vpicone** [comment]
@asudoh removing the height from dropdown and instead using the listbox styles solves the problem.

We need an exception for vanilla which does not use listbox:

```
  .#{$prefix}--dropdown:not(.#{$prefix}--list-box) {
    height: rem(40px);
  }
```

This approach means only vanilla (and even then only the dropdown) is impacted by extra specificity.

I believe this is a better approach compared to doubling the specificity for all frameworks and for all three list box components. At this point, we shouldn't be sacrificing style quality for Vanilla. Since this is a new feature, we could have made the Vanilla implementation require `listbox` classes to function (or even better, not introduce these variants for vanilla at all)

**Comment by asudoh** [comment]
Just in case it clarifies, the list box classes cannot be applies to the legacy markup due to the nature of difference in markup.

**Review by emyarod — COMMENTED** [review]
it looks like [`.bx--dropdown--xl` is already in the stylesheet](https://github.com/carbon-design-system/carbon/blob/master/packages/components/src/components/dropdown/_dropdown.scss#L69-L71), it is just not being added to the React component. since multiselect and combobox are not present in the vanilla library and dropdown does not use listbox classes in vanilla, should we just apply the `.bx--dropdown--xl` class to the React component for now?

**Comment by asudoh** [comment]
@emyarod Sounds 💯, now I wonder why I didn't come up with such idea... Updated.

File: packages/components/src/components/list-box/_list-box.scss [hunk 1]
```diff
@@ -86,12 +86,12 @@ $list-box-menu-width: rem(300px);
     }
   }
 
-  .#{$prefix}--list-box--xl {
+  #{$prefix}--list-box--xl {
```
Comment (emyarod) [inline packages/components/src/components/list-box/_list-box.scss]:
```suggestion
  .#{$prefix}--list-box--xl {
```

File: packages/components/src/components/list-box/_list-box.scss [hunk 2]
```diff
@@ -86,12 +86,12 @@ $list-box-menu-width: rem(300px);
...
     height: rem(48px);
     max-height: rem(48px);
   }
 
-  .#{$prefix}--list-box--sm {
+  #{$prefix}--list-box--sm {
```
Comment (emyarod) [inline packages/components/src/components/list-box/_list-box.scss]:
```suggestion
  .#{$prefix}--list-box--sm {
```

**Review by emyarod — COMMENTED** [review]
was the selector change intentional?

**Comment by asudoh** [comment]
Oops good catch @emyarod! Not sure why the wrong selectors got into the commit...

**Review by joshblack — APPROVED** [review]

**Review by emyarod — APPROVED** [review]
thanks for making the change, looks good to me

**Comment by vpicone** [comment]
@joshblack @emyarod those drop down selectors were added with @asudoh’s original PR last month. They went unused in that PR (#4731)

The size functionality in react got randomly lumped in with this unrelated PR: #4999. Now this PR just wires it up?

Isn’t weird the other listbox components don’t need their own size variant classes? Or that we're still passing the size into the Listbox of dropdown?

Also the arrow is misaligned for the size variants in iOS: 
DFC93455-0F75-40D0-A1E4-05338CA0CC11

**Comment by joshblack** [comment]
@vpicone oh wow, did not realize this. How can we address this? Do we need to revert this PR?

**Comment by vpicone** [comment]
@joshblack this was my recommendation and has context/explanation in the comments #4917 

The reality is our vanilla implementation of dropdown differs from react (it doesn’t use listbox). So we have to choose the “least wrong” way to style these variants. That PR has a bunch of info on why I thought that approach was the lesser of two evils. Optimizing for parity with other listbox components and minimizing specificity. 

The issue with reverting this PR is that the actual substance of the implementation (wiring the selectors to the react component) is in an unrelated PR: #4999 . That PR likely has changes we don’t want to revert.

**Comment by asudoh** [comment]
`.bx--dropdown--xl` is actually something for use - It was introduced so that size variants can be used in legacy markup.

**Comment by vpicone** [comment]
@asudoh So we're okay with including classes/styles for particular frameworks now?

You've now made it the implementation for React as well as legacy markup. Now React Dropdown get's size variant from both Listbox and this new Dropdown class you've added.

**Comment by asudoh** [comment]
The legacy markup is not for particular frameworks, just a left-over from earlier release. That said, usage of `.bx--dropdown--xl` class is an interim solution until we can make a breaking change.

**Comment by vpicone** [comment]
@asudoh legacy markup is vanilla no? The Dropdown is now styled by both these new classes you added for the “legacy system” _and_ listbox classes.

**Comment by asudoh** [comment]
As stated earlier, cleaning up legacy code can happen only when we can make a breaking change.

**Comment by vpicone** [comment]
@asudoh but this isn’t legacy it’s a new feature. You just added these selectors: https://github.com/asudoh/carbon-components/commit/0723836da5ebcadd1c8da689483ce159a8fd81a7#diff-d69d1c6c9ae9007ad44b45f8f333fa3f

Were they only meant for vanilla?

**Comment by asudoh** [comment]
New feature, while given it's in `v10` codebase, compatibility with existing codebase in `v10` needs to be taken into account.

**Pull request merged**